File: app/src/main/java/com/cooper/wheellog/utils/VeteranAdapter.java
package com.cooper.wheellog.utils;

import com.cooper.wheellog.WheelData;
import com.cooper.wheellog.WheelLog;

import java.io.ByteArrayOutputStream;
import java.util.Locale;
import java.util.zip.CRC32;
import timber.log.Timber;

public class VeteranAdapter extends BaseAdapter {
    private static VeteranAdapter INSTANCE;
    veteranUnpacker unpacker = new veteranUnpacker();
    private static final int WAITING_TIME = 100;
    private long time_old = 0;
    private int mVer = 0;

    @Override
    public boolean decode(byte[] data) {
        Timber.i("Decode Veteran");
        WheelData wd = WheelData.getInstance();
        wd.resetRideTime();
        long time_new = System.currentTimeMillis();
        if ((time_new-time_old) > WAITING_TIME) // need to reset state in case of packet loose
            unpacker.reset();
        time_old = time_new;
        boolean newDataFound = false;
        for (byte c : data) {
            if (unpacker.addChar(c)) {
                byte[] buff = unpacker.getBuffer();
                Boolean useBetterPercents = WheelLog.AppConfig.getUseBetterPercents();
                Boolean hwPwmEnabled = WheelLog.AppConfig.getHwPwm();
                int veteranNegative = Integer.parseInt(WheelLog.AppConfig.getGotwayNegative());
                int voltage = MathsUtil.shortFromBytesBE(buff,4);
                int speed = MathsUtil.signedShortFromBytesBE(buff,6) * 10;
                int distance = MathsUtil.intFromBytesRevBE(buff,8);
                int totalDistance = MathsUtil.intFromBytesRevBE(buff, 12);
                int phaseCurrent = MathsUtil.signedShortFromBytesBE(buff,16) * 10;
                int temperature = MathsUtil.signedShortFromBytesBE(buff, 18);
                int autoOffSec = MathsUtil.shortFromBytesBE(buff,20);
                int chargeMode = MathsUtil.shortFromBytesBE(buff,22);
                int speedAlert = MathsUtil.shortFromBytesBE(buff, 24) * 10;
                int speedTiltback = MathsUtil.shortFromBytesBE(buff,26) * 10;
                int ver = MathsUtil.shortFromBytesBE(buff,28);
                mVer = ver/1000;
                String version = String.format(Locale.US, "%03d.%01d.%02d", ver/1000, (ver%1000)/100, (ver%100));
                int pedalsMode = MathsUtil.shortFromBytesBE(buff, 30);
                int pitchAngle = MathsUtil.signedShortFromBytesBE(buff,32);
                int hwPwm = MathsUtil.shortFromBytesBE(buff, 34);

                // smartBMS part
                if (mVer >= 5) { // Lynx = 5; Sherman L = 6
                    int pnum = buff[46];
                    int bmsnum = pnum < 4 ? 1 : 2;
                    SmartBms bms = bmsnum == 1 ? wd.getBms1() : wd.getBms2();

                    if (pnum == 0 || pnum == 4) {
                        /*
                        System.out.print(String.format(Locale.US, "Unknown: "));
                        for (int i = 0; i < 13; i++) {
                            int cell1 = MathsUtil.signedShortFromBytesBE(buff, 47 + i * 2);
                            System.out.print(String.format(Locale.US, " %d: %d,", i, cell1));
                        }
                        System.out.print(String.format(Locale.US, "\n"));
                        */
                        wd.getBms1().setCurrent((double) MathsUtil.signedShortFromBytesBE(buff, 69)/100.0);
                        wd.getBms2().setCurrent((double) MathsUtil.signedShortFromBytesBE(buff, 71)/100.0);
                    } else if (pnum == 1 | pnum == 5) {
                        for (int i = 0; i < 15; i++) {
                            int cell = MathsUtil.signedShortFromBytesBE(buff, 53 + i * 2);
                            bms.getCells()[i] = cell/1000.0;
                        }
                    } else if (pnum == 2 | pnum == 6) {
                        for (int i = 0; i < 15; i++) {
                            int cell = MathsUtil.shortFromBytesBE(buff, 53 + i * 2);
                            bms.getCells()[i+15] = cell/1000.0;
                        }
                    } else if (pnum == 3 | pnum == 7) {
                        for (int i = 0; i < 6; i++) {
                            int cell = MathsUtil.shortFromBytesBE(buff, 59 + i * 2);
                            bms.getCells()[i+30] = cell/1000.0;
                        }
                        bms.setTemp1(MathsUtil.shortFromBytesBE(buff, 47)/100.0);
                        bms.setTemp2(MathsUtil.shortFromBytesBE(buff, 49)/100.0);
                        bms.setTemp3(MathsUtil.shortFromBytesBE(buff, 51)/100.0);
                        bms.setTemp4(MathsUtil.shortFromBytesBE(buff, 53)/100.0);
                        bms.setTemp5(MathsUtil.shortFromBytesBE(buff, 55)/100.0);
                        bms.setTemp6(MathsUtil.shortFromBytesBE(buff, 57)/100.0);

                        bms.setMinCell(bms.getCells()[0]);
                        bms.setMaxCell(bms.getCells()[0]);
                        double totalVolt = 0.0;
                        for (int i = 0; i < getCellsForWheel(); i++) {
                            double cell = bms.getCells()[i];
                            totalVolt += cell;
                            if (cell > 0.0) {
                                if (bms.getMaxCell() < cell) {
                                    bms.setMaxCell(cell);
                                }
                                if (bms.getMinCell() > cell) {
                                    bms.setMinCell(cell);
                                }
                            }
                        }
                        bms.setCellDiff(bms.getMaxCell() - bms.getMinCell());
                        bms.setVoltage(totalVolt);
                    }
                }
                // end of smartBMS part

                int battery;
                if (mVer < 4) { // not Patton
                    if (useBetterPercents) {
                        if (voltage > 10020) {
                            battery = 100;
                        } else if (voltage > 8160) {
                            battery = (int) Math.round((voltage - 8070) / 19.5);
                        } else if (voltage > 7935) {
                            battery = (int) Math.round((voltage - 7935) / 48.75);
                        } else {
                            battery = 0;
                        }
                    } else {
                        if (voltage <= 7935) {
                            battery = 0;
                        } else if (voltage >= 9870) {
                            battery = 100;
                        } else {
                            battery = (int) Math.round((voltage - 7935) / 19.5);
                        }
                    }
                } else if (mVer == 4) { // Patton
                    if (useBetterPercents) {
                        if (voltage > 12525) {
                            battery = 100;
                        } else if (voltage > 10200) {
                            battery = (int) Math.round((voltage - 9975) / 25.5);
                        } else if (voltage > 9600) {
                            battery = (int) Math.round((voltage - 9600) / 67.5);
                        } else {
                            battery = 0;
                        }
                    } else {
                        if (voltage <= 9918) {
                            battery = 0;
                        } else if (voltage >= 12337) {
                            battery = 100;
                        } else {
                            battery = (int) Math.round((voltage - 9918) / 24.2);
                        }
                    }
                } else if (mVer >= 5) { // Lynx = 5; Sherman L = 6
                    if (useBetterPercents) {
                        if (voltage > 15030) {
                            battery = 100;
                        } else if (voltage > 12240) {
                            battery = (int) Math.round((voltage - 11970) / 30.6);
                        } else if (voltage > 11520) {
                            battery = (int) Math.round((voltage - 11520) / 81);
                        } else {
                            battery = 0;
                        }
                    } else {
                        if (voltage <= 11902) {
                            battery = 0;
                        } else if (voltage >= 14805) {
                            battery = 100;
                        } else {
                            battery = (int) Math.round((voltage - 11902) / 29.03);
                        }
                    }
                } else battery = 1; // for new wheels, set 1% by default

                if (veteranNegative == 0) {
                    speed = Math.abs(speed);
                    phaseCurrent = Math.abs(phaseCurrent);
                } else {
                    speed = speed * veteranNegative;
                    phaseCurrent = phaseCurrent * veteranNegative;
                }

                wd.setVersion(version);
                wd.setSpeed(speed);
                wd.setTopSpeed(speed);
                wd.setWheelDistance(distance);
                wd.setTotalDistance(totalDistance);
                wd.setTemperature(temperature);
                wd.setPhaseCurrent(phaseCurrent);
                wd.setVoltage(voltage);
                wd.setBatteryLevel(battery);
                wd.setChargingStatus(chargeMode);
                wd.setAngle(pitchAngle/100.0);
                if (hwPwmEnabled) {
                    wd.setOutput(hwPwm);
                    wd.updatePwm();
                } else {
                    wd.calculatePwm();
                }
                wd.calculateCurrent();
                wd.calculatePower();
                wd.setModel(getModel());
                newDataFound = true;
            }
        }
        return newDataFound;
    }

    @Override
    public boolean isReady() {
        return WheelData.getInstance().getVoltage() != 0 && mVer != 0;
    }

    public void resetTrip() {
        WheelData.getInstance().bluetoothCmd("CLEARMETER".getBytes());
    }

    @Override
    public void updatePedalsMode(int pedalsMode) {
        switch (pedalsMode) {
            case 0:
                WheelData.getInstance().bluetoothCmd("SETh".getBytes());
                break;
            case 1:
                WheelData.getInstance().bluetoothCmd("SETm".getBytes());
                break;
            case 2:
                WheelData.getInstance().bluetoothCmd("SETs".getBytes());
                break;
        }
    }
    public int getVer() {
        if (mVer >=2) {
            WheelLog.AppConfig.setHwPwm(true);
        }
        return mVer;
    }

    private String getModel() {
        String vModel = "";
        if (mVer <= 1) vModel = "Sherman"; else
        if (mVer == 2) vModel = "Abrams"; else
        if (mVer == 3) vModel = "Sherman S"; else
        if (mVer == 4) vModel = "Patton"; else
        if (mVer == 5) vModel = "Lynx"; else
        if (mVer == 6) vModel = "Sherman L"; else vModel = "Unknown";
        return vModel;
    }

    @Override
    public void switchFlashlight() {
        boolean light = !WheelLog.AppConfig.getLightEnabled();
        WheelLog.AppConfig.setLightEnabled(light);
        setLightState(light);
    }

    @Override
    public void setLightState(final boolean lightEnable) {
        String command = "";
        if (lightEnable) {
            command = "SetLightON";
        } else {
            command = "SetLightOFF";
        }
        WheelData.getInstance().bluetoothCmd(command.getBytes());
    }

    @Override
    public int getCellsForWheel() {
        if (mVer >= 5) {
            return 36;
        } else if (mVer == 4) {
            return 30;
        } else {
            return 24;
        }
    }

    @Override
    public void wheelBeep() {
        if (mVer < 3) {
            WheelData.getInstance().bluetoothCmd("b".getBytes());
        } else {
            WheelData.getInstance().bluetoothCmd(new byte[]{ 0x4c, 0x6b, 0x41, 0x70, 0x0e, 0x00, (byte) 0x80,(byte) 0x80,(byte) 0x80, 0x01,(byte) 0xca,(byte) 0x87,(byte) 0xe6, 0x6f});
        }

    }

    static class veteranUnpacker {
        private boolean usingCrc = false;
        enum UnpackerState {
            unknown,
            collecting,
            lensearch,
            done
        }


        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        int old1 = 0;
        int old2 = 0;
        int len = 0;

        UnpackerState state = UnpackerState.unknown;

        byte[] getBuffer() {
            return buffer.toByteArray();
        }

        boolean addChar(int c) {

            switch (state) {

                case collecting:

                    int bsize = buffer.size();
                    if (((bsize == 22 || bsize == 30) && (c != 0x00)) || ((bsize == 23) && ((c & 0xFE) != 0x00)) ) {
                        state = UnpackerState.done;
                        Timber.i("Data verification failed");
                        reset();
                        return false;
                    }
                    buffer.write(c);
                    if (bsize == len+3) {
                        state = UnpackerState.done;
                        Timber.i("Len %d", len);
                        Timber.i("Step reset");
                        reset();
                        if (len > 38 || usingCrc) { // new format with crc32
                            CRC32 crc = new CRC32();
                            crc.update(getBuffer(), 0, len);
                            long calc_crc = crc.getValue();
                            long provided_crc = MathsUtil.intFromBytesBE(getBuffer(), len);
                            if (calc_crc == provided_crc) {
                                usingCrc = true;
                                Timber.i("CRC32 ok");
                                return true;
                            } else {
                                Timber.i("CRC32 fail");
                                return false;
                            }
                        }
                        return true; // old format without crc32

                    }
                    break;

                case lensearch:
                    buffer.write(c);
                    len = c & 0xff;
                    state = UnpackerState.collecting;
                    old2 = old1;
                    old1 = c;
                    break;


                default:
                    if (c == (byte) 0x5C && old1 == (byte) 0x5A && old2 == (byte) 0xDC) {
                        buffer = new ByteArrayOutputStream();
                        buffer.write(0xDC);
                        buffer.write(0x5A);
                        buffer.write(0x5C);
                        state = UnpackerState.lensearch;
                    } else if (c == (byte) 0x5A && old1 == (byte) 0xDC) {
                        old2 = old1;
                    } else {
                        old2 = 0;
                    }
                    old1 = c;

            }
            return false;
        }

        void reset() {
            old1 = 0;
            old2 = 0;
            state = UnpackerState.unknown;

        }
    }


    public static VeteranAdapter getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new VeteranAdapter();
        }
        return INSTANCE;
    }

}

File: app/src/main/java/com/cooper/wheellog/utils/VeteranAdapter.kt
package com.cooper.wheellog.utils

import com.cooper.wheellog.WheelData
import com.cooper.wheellog.WheelLog
import timber.log.Timber
import java.io.ByteArrayOutputStream
import java.util.Locale
import java.util.zip.CRC32
import kotlin.math.abs
import kotlin.math.roundToInt

class VeteranAdapter : BaseAdapter() {
    private var unpacker = VeteranUnpacker()
    private var oldTime: Long = 0
    private var manufacturerVersion = 0
    override fun decode(data: ByteArray?): Boolean {
        Timber.i("Decode Veteran")
        val wd = WheelData.getInstance()
        wd.resetRideTime()
        val newTime = System.currentTimeMillis()
        if (newTime - oldTime > WAITING_TIME) {
            // need to reset state in case of packet loose
            unpacker.reset()
        }
        oldTime = newTime
        var newDataFound = false
        for (c in data!!) {
            if (unpacker.addChar(c.toInt())) {
                val buff = unpacker.getBuffer()
                val useBetterPercents = WheelLog.AppConfig.useBetterPercents
                val veteranNegative = WheelLog.AppConfig.gotwayNegative.toInt()
                val voltage = MathsUtil.shortFromBytesBE(buff, 4)
                var speed = MathsUtil.signedShortFromBytesBE(buff, 6) * 10
                val distance = MathsUtil.intFromBytesRevBE(buff, 8)
                val totalDistance = MathsUtil.intFromBytesRevBE(buff, 12)
                var phaseCurrent = MathsUtil.signedShortFromBytesBE(buff, 16) * 10
                val temperature = MathsUtil.signedShortFromBytesBE(buff, 18)
                val autoOffSec = MathsUtil.shortFromBytesBE(buff, 20)
                val chargeMode = MathsUtil.shortFromBytesBE(buff, 22)
                val speedAlert = MathsUtil.shortFromBytesBE(buff, 24) * 10
                val speedTiltback = MathsUtil.shortFromBytesBE(buff, 26) * 10
                val ver = MathsUtil.shortFromBytesBE(buff, 28)
                manufacturerVersion = ver / 1000
                val version = String.format(
                    Locale.US,
                    "%03d.%01d.%02d",
                    ver / 1000,
                    ver % 1000 / 100,
                    ver % 100,
                )
                val pedalsMode = MathsUtil.shortFromBytesBE(buff, 30)
                val pitchAngle = MathsUtil.signedShortFromBytesBE(buff, 32)
                val hwPwm = MathsUtil.shortFromBytesBE(buff, 34)
                val battery: Int = computeBatteryPercentage(useBetterPercents, voltage)
                if (veteranNegative == 0) {
                    speed = abs(speed)
                    phaseCurrent = abs(phaseCurrent)
                } else {
                    speed *= veteranNegative
                    phaseCurrent *= veteranNegative
                }
                wd.version = version
                wd.speed = speed
                wd.topSpeed = speed
                wd.setWheelDistance(distance.toLong())
                wd.totalDistance = totalDistance.toLong()
                wd.temperature = temperature
                wd.phaseCurrent = phaseCurrent
                wd.current = phaseCurrent
                wd.voltage = voltage
                wd.voltageSag = voltage
                wd.batteryLevel = battery
                wd.chargingStatus = chargeMode
                wd.angle = pitchAngle / 100.0
                wd.output = hwPwm
                wd.updateRideTime()
                newDataFound = true
            }
        }
        return newDataFound
    }

    /**
     * Compute battery percentage from voltage
     * @param useBetterPercents use better percents is a flag set by the user
     * @param voltage voltage in mV
     * TODO correct the voltage values to work for the LeaperKim Lynx and other high voltage wheels
     */
    private fun computeBatteryPercentage(useBetterPercents: Boolean, voltage: Int): Int {
        return if (manufacturerVersion < 4) {
            // For non patton wheels, the voltage is 0-100V. This will change with the LeaperKim Lynx.
            if (useBetterPercents) {
                when {
                    voltage > SHERMAN_S_MAX_VOLTAGE -> 100
                    voltage > 8160 -> ((voltage - 8070) / 19.5).roundToInt()
                    voltage > MIN_VOLTAGE -> ((voltage - 7935) / 48.75).roundToInt()
                    else -> 0
                }
            } else {
                when {
                    voltage <= MIN_VOLTAGE -> 0
                    voltage >= 9870 -> 100
                    else -> ((voltage - 7935) / 19.5).roundToInt()
                }
            }
        } else {
            // For patton wheels, the voltage is 0-126V
            if (useBetterPercents) {
                when {
                    voltage > PATTON_MAX_VOLTAGE_BETTER_PERCENT -> 100
                    voltage > SHERMAN_S_MAX_VOLTAGE -> ((voltage - 9975) / 25.5).roundToInt()
                    voltage > 9600 -> ((voltage - 9600) / 67.5).roundToInt()
                    else -> 0
                }
            } else {
                when {
                    voltage <= 9918 -> 0
                    voltage >= PATTON_MAX_VOLTAGE -> 100
                    else -> ((voltage - 9918) / 24.2).roundToInt()
                }
            }
        }
    }

    override val isReady: Boolean
        get() = WheelData.getInstance().voltage != 0 && manufacturerVersion != 0

    fun resetTrip() {
        WheelData.getInstance().bluetoothCmd("CLEARMETER".toByteArray())
    }

    override fun updatePedalsMode(pedalsMode: Int) {
        when (pedalsMode) {
            0 -> WheelData.getInstance().bluetoothCmd("SETh".toByteArray())
            1 -> WheelData.getInstance().bluetoothCmd("SETm".toByteArray())
            2 -> WheelData.getInstance().bluetoothCmd("SETs".toByteArray())
        }
    }

    val ver: Int
        get() {
            if (manufacturerVersion >= 2) {
                WheelLog.AppConfig.hwPwm = true
            }
            return manufacturerVersion
        }

    override fun switchFlashlight() {
        val light = !WheelLog.AppConfig.lightEnabled
        WheelLog.AppConfig.lightEnabled = light
        setLightState(light)
    }

    override fun setLightState(on: Boolean) {
        var command = ""
        command = if (on) {
            "SetLightON"
        } else {
            "SetLightOFF"
        }
        WheelData.getInstance().bluetoothCmd(command.toByteArray())
    }

    /**
     * TODO Cells for wheel is not necessarily correct for the LeaperKim Lynx
     */
    override val cellsForWheel: Int
        get() = if (manufacturerVersion > 3) {
            30
        } else {
            24
        }

    override fun wheelBeep() {
        WheelData.getInstance().bluetoothCmd("b".toByteArray())
    }

    private class VeteranUnpacker {
        enum class UnpackerState {
            UNKNOWN, COLLECTING, LENS_SEARCH, DONE
        }

        var buffer = ByteArrayOutputStream()
        var old1 = 0
        var old2 = 0
        var len = 0
        var state = UnpackerState.UNKNOWN
        fun getBuffer(): ByteArray {
            return buffer.toByteArray()
        }

        fun addChar(c: Int): Boolean {
            return when (state) {
                UnpackerState.COLLECTING -> handleCollectingState(c)
                UnpackerState.LENS_SEARCH -> handleLensSearchState(c)
                UnpackerState.UNKNOWN -> handleUnknownOrDoneState(c)
                UnpackerState.DONE -> handleUnknownOrDoneState(c)
            }
        }

        private fun handleUnknownOrDoneState(c: Int): Boolean {
            when {
                c == 0x5C.toByte().toInt() && old1 == 0x5A.toByte()
                    .toInt() && old2 == 0xDC.toByte().toInt() -> {
                    buffer = ByteArrayOutputStream()
                    buffer.write(0xDC)
                    buffer.write(0x5A)
                    buffer.write(0x5C)
                    state = UnpackerState.LENS_SEARCH
                }

                c == 0x5A.toByte().toInt() && old1 == 0xDC.toByte().toInt() -> {
                    old2 = old1
                }

                else -> {
                    old2 = 0
                }
            }
            old1 = c
            return false
        }

        private fun handleLensSearchState(c: Int): Boolean {
            buffer.write(c)
            len = c and 0xff
            state = UnpackerState.COLLECTING
            old2 = old1
            old1 = c
            return false
        }

        private fun handleCollectingState(c: Int): Boolean {
            val bufferSize = buffer.size()
            return if (shouldTransitionToDone(bufferSize, c)) {
                state = UnpackerState.DONE
                Timber.i("Data verification failed")
                reset()
                false
            } else {
                buffer.write(c)
                if (bufferSize == len + 3) {
                    state = UnpackerState.DONE
                    Timber.i("Len %d", len)
                    Timber.i("Step reset")
                    reset()
                    if (len > 38) {
                        crcFormatParsing(
                            len,
                            getBuffer(),
                        )
                    } else {
                        true // old format without crc32
                    }
                } else {
                    false
                }
            }
        }

        private fun shouldTransitionToDone(bufferSize: Int, c: Int) =
            (bufferSize == 22 || bufferSize == 30) && c != 0x00 || bufferSize == 23 && c and 0xFE != 0x00 || bufferSize == 31 && c and 0xFC != 0x00

        /**
         * CRC32 format parsing
         * @param length length of the buffer
         * @return true if crc32 is ok
         * @return false if crc32 is wrong
         * TODO send failures to some sort of analytics/tracking service
         */
        private fun crcFormatParsing(length: Int, buffer: ByteArray): Boolean {
            val crc = CRC32()
            try {
                crc.update(buffer, 0, length - 4)
            } catch (e: ArrayIndexOutOfBoundsException) {
                Timber.i("CRC32 fail")
                Timber.e(e)
                return false
            } catch (e: NullPointerException) {
                Timber.i("CRC32 fail")
                Timber.e(e)
                return false
            }
            val calculatedCrc = crc.value
            val providedCrc = MathsUtil.getInt4(buffer, length)
            return if (calculatedCrc == providedCrc) {
                Timber.i("CRC32 ok")
                true
            } else {
                Timber.i("CRC32 fail")
                false
            }
        }

        fun reset() {
            old1 = 0
            old2 = 0
            state = UnpackerState.UNKNOWN
        }
    }

    companion object {
        private var INSTANCE: VeteranAdapter? = null
        private const val WAITING_TIME = 100

        // The minimum value that the wheel will report as a voltage in bytebuffer
        private const val MIN_VOLTAGE = 7935
        private const val SHERMAN_S_MAX_VOLTAGE = 10020

        // The maximum value that the wheel will report as a voltage in bytebuffer
        private const val PATTON_MAX_VOLTAGE_BETTER_PERCENT = 12525
        private const val PATTON_MAX_VOLTAGE = 12337

        @JvmStatic
        val instance: VeteranAdapter
            get() {
                if (INSTANCE == null) {
                    INSTANCE = VeteranAdapter()
                }
                return INSTANCE!!
            }
    }
}
